File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        res = []
        
        if digits == "":
            return res
         
        dict1 = {2 : "abc", 3 : "def", 4 : "ghi", 5 : "jkl", 6 : "mno", 7 : "pqrs", 8 : "tuv", 9 : "wxyz"}
        
        
        letter_string = dict1[int(digits[0])]   
        
        if len(digits) > 1:
            sub_output = self.letterCombinations(digits[1:])

            for c in letter_string:
                for output in sub_output:
                    res.append(c + output)
            
            return res
        
        else:
            return [c for c in letter_string]

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
import java.util.*;

public class Solution {
    public List<String> letterCombinations(String digits) {
      List<String> res = new ArrayList<>();
      if (digits.equals(""))
        return res;

      Map<Integer, String> dict1 = new HashMap<>();
      dict1.put(2, "abc");
      dict1.put(3, "def");
      dict1.put(4, "ghi");
      dict1.put(5, "jkl");
      dict1.put(6, "mno");
      dict1.put(7, "pqrs");
      dict1.put(8, "tuv");
      dict1.put(9, "wxyz");

      String letter_string = dict1.get(Integer.parseInt(digits.substring(0, 1)));

      if (digits.length() > 1) {
        List<String> sub_output = letterCombinations(digits.substring(1));
        for (char c : letter_string.toCharArray()) {
          for (String output : sub_output) {
            res.add(c + output);
          }
        }
        return res;
      }
      else {
        for (char c : letter_string.toCharArray()) {
          res.add(String.valueOf(c));
        }
        return res;
      }
    }
}